### PR TITLE
[TTAHUB-853] New AR topics migration

### PR DESCRIPTION
**Pull request by nvms**
## After `up`

Old topics should correctly map to the new topic and have good `deletedAt` values:

Screen Shot 2022-08-17 at 10 42 33 AM

Screen Shot 2022-08-17 at 10 42 40 AM

The new topic name should be replaced in the `topics` column in `ActivityReports` table:

Screen Shot 2022-08-17 at 10 43 28 AM

## After `down`

`topics` column in `ActivityReports` table reverted to original values:

Screen Shot 2022-08-17 at 10 44 06 AM

`mapsTo` and `deletedAt` restored to original values in `ActivityReports` table:

Screen Shot 2022-08-17 at 10 44 20 AM

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-853

### Every PR

- [x] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/migrations/20220812205424-new-ar-topics.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+          ('Ongoing Monitoring Management System', current_timestamp, current_timestamp),
+          ('Training and Professional Development', current_timestamp, current_timestamp);
+        `,
+        { transaction },
+      );
+
+      // --------------------------------------------
+      // Update `mapsTo` field of existing topics to point to new topics
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" t1 SET "mapsTo" = t2.id, "deletedAt" = current_timestamp FROM "Topics" t2 WHERE t1.name = \'Child Assessment, Development, Screening\' AND t2.name = \'Child Screening and Assessment\' AND t1."deletedAt" IS NULL;',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:48]:
```suggestion
        `UPDATE "Topics" t1 
        SET 
            "mapsTo" = t2.id, 
            "deletedAt" = current_timestamp 
        FROM "Topics" t2 
        WHERE t1.name = 'Child Assessment, Development, Screening' 
        AND t2.name = 'Child Screening and Assessment' 
        AND t1."deletedAt" IS NULL;`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        { transaction },
+      );
+
+      // Update `mapsTo` in topics for all topics that share the same old name - does NOT update deletedAt (would violate constraint [name-deletedAt-mapsTo])
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" t1 SET "mapsTo" = t2.id FROM "Topics" t2 WHERE t1.name = \'Child Assessment, Development, Screening\' AND t2.name = \'Child Screening and Assessment\' AND t2."deletedAt" IS NULL;',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:54]:
```suggestion
        `UPDATE "Topics" t1 
        SET 
            "mapsTo" = t2.id 
        FROM "Topics" t2 
        WHERE t1.name = 'Child Assessment, Development, Screening' 
        AND t2.name = 'Child Screening and Assessment' 
        AND t2."deletedAt" IS NULL;`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        { transaction },
+      );
+
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" t1 SET "mapsTo" = t2.id, "deletedAt" = current_timestamp FROM "Topics" t2 WHERE t1.name = \'Teaching Practices / Teacher-Child Interactions\' AND t2.name = \'Teaching / Caregiving Practices\' AND t1."deletedAt" IS NULL;',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:59]:
```suggestion
        `UPDATE "Topics" t1 
        SET 
            "mapsTo" = t2.id, 
            "deletedAt" = current_timestamp 
        FROM "Topics" t2 
        WHERE t1.name = 'Teaching Practices / Teacher-Child Interactions' 
        AND t2.name = 'Teaching / Caregiving Practices' 
        AND t1."deletedAt" IS NULL;`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+      // --------------------------------------------
+      // Update `topics` column of existing ActivityReport records to use the new topics:
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Child Assessment, Development, Screening\', \'Child Screening and Assessment\') WHERE topics @> \'{"Child Assessment, Development, Screening"}\';',
+        { transaction },
+      );
+
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Teaching Practices / Teacher-Child Interactions\', \'Teaching / Caregiving Practices\') WHERE topics @> \'{"Teaching Practices / Teacher-Child Interactions"}\';',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:76]:
```suggestion
        `UPDATE "ActivityReports" 
        SET
            topics = array_replace(topics, 'Teaching Practices / Teacher-Child Interactions', 'Teaching / Caregiving Practices') 
        WHERE topics @> '{"Teaching Practices / Teacher-Child Interactions"}';`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        { transaction },
+      );
+
+      // Now, any deleted topics that share this name should map to this undeleted topic
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" t1 SET "mapsTo" = t2.id FROM "Topics" t2 WHERE t1.name = \'Child Assessment, Development, Screening\' AND t1."deletedAt" IS NOT NULL AND t2.name = \'Child Assessment, Development, Screening\' AND t2."deletedAt" is NULL;',
+        { transaction },
+      );
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" t1 SET "mapsTo" = t2.id FROM "Topics" t2 WHERE t1.name = \'Teaching Practices / Teacher-Child Interactions\' AND t1."deletedAt" IS NOT NULL AND t2.name = \'Teaching Practices / Teacher-Child Interactions\' AND t2."deletedAt" is NULL;',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:145]:
```suggestion
        `UPDATE "Topics" t1 
        SET
            "mapsTo" = t2.id 
        FROM "Topics" t2 
        WHERE t1.name = 'Teaching Practices / Teacher-Child Interactions' 
        AND t1."deletedAt" IS NOT NULL 
        AND t2.name = 'Teaching Practices / Teacher-Child Interactions' 
        AND t2."deletedAt" is NULL;`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        { transaction },
+      );
+
+      // "Un-delete" the recently deleted topics.
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" SET "mapsTo" = null, "deletedAt" = null WHERE name = \'Child Assessment, Development, Screening\' AND "deletedAt" = (SELECT max("deletedAt") FROM "Topics" WHERE name = \'Child Assessment, Development, Screening\');',
+        { transaction },
+      );
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" SET "mapsTo" = null, "deletedAt" = null WHERE name = \'Teaching Practices / Teacher-Child Interactions\' AND "deletedAt" = (SELECT max("deletedAt") FROM "Topics" WHERE name = \'Teaching Practices / Teacher-Child Interactions\');',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:135]:
```suggestion
        `UPDATE "Topics" 
        SET
            "mapsTo" = null,
            "deletedAt" = null 
        WHERE name = 'Teaching Practices / Teacher-Child Interactions' 
        AND "deletedAt" = (
            SELECT max("deletedAt") 
            FROM "Topics" 
            WHERE name = 'Teaching Practices / Teacher-Child Interactions');`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      );
+
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Teaching / Caregiving Practices\', \'Teaching Practices / Teacher-Child Interactions\') WHERE topics @> \'{"Teaching / Caregiving Practices"}\';',
+        { transaction },
+      );
+
+      // "Un-delete" the recently deleted topics.
+      await queryInterface.sequelize.query(
+        'UPDATE "Topics" SET "mapsTo" = null, "deletedAt" = null WHERE name = \'Child Assessment, Development, Screening\' AND "deletedAt" = (SELECT max("deletedAt") FROM "Topics" WHERE name = \'Child Assessment, Development, Screening\');',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:131]:
```suggestion
        `UPDATE "Topics" 
        SET 
            "mapsTo" = null,
            "deletedAt" = null 
        WHERE name = 'Child Assessment, Development, Screening' 
        AND "deletedAt" = (
            SELECT max("deletedAt") 
            FROM "Topics" 
            WHERE name = 'Child Assessment, Development, Screening');`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+      );
+
+      // Revert the values in the topics array column of ActivityReports to their previous values.
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Child Screening and Assessment\', \'Child Assessment, Development, Screening\') WHERE topics @> \'{"Child Screening and Assessment"}\';',
+        { transaction },
+      );
+
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Teaching / Caregiving Practices\', \'Teaching Practices / Teacher-Child Interactions\') WHERE topics @> \'{"Teaching / Caregiving Practices"}\';',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:125]:
```suggestion
        `UPDATE "ActivityReports" 
        SET 
            topics = array_replace(topics, 'Teaching / Caregiving Practices', 'Teaching Practices / Teacher-Child Interactions') 
        WHERE topics @> '{"Teaching / Caregiving Practices"}';`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+      // Delete all the new topics
+      await queryInterface.sequelize.query(
+        'DELETE FROM "Topics" WHERE name IN (\'Child Screening and Assessment\', \'Teaching / Caregiving Practices\', \'Disabilities Services\', \'Ongoing Monitoring Management System\', \'Training and Professional Development\') AND "deletedAt" IS NULL;',
+        { transaction },
+      );
+
+      // Revert the values in the topics array column of ActivityReports to their previous values.
+      await queryInterface.sequelize.query(
+        'UPDATE "ActivityReports" SET topics = array_replace(topics, \'Child Screening and Assessment\', \'Child Assessment, Development, Screening\') WHERE topics @> \'{"Child Screening and Assessment"}\';',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:120]:
```suggestion
        `UPDATE "ActivityReports" .
        SET 
            topics = array_replace(topics, 'Child Screening and Assessment', 'Child Assessment, Development, Screening') 
        WHERE topics @> '{"Child Screening and Assessment"}';`,
```

File: src/migrations/20220812205424-new-ar-topics.js [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+      await queryInterface.sequelize.query(
+        `
+          SELECT "ZAFSetTriggerState"(null, null, null, 'DISABLE');
+          `,
+        { transaction },
+      );
+
+      // Delete all the new topics
+      await queryInterface.sequelize.query(
+        'DELETE FROM "Topics" WHERE name IN (\'Child Screening and Assessment\', \'Teaching / Caregiving Practices\', \'Disabilities Services\', \'Ongoing Monitoring Management System\', \'Training and Professional Development\') AND "deletedAt" IS NULL;',
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:114]:
```suggestion
        `DELETE FROM "Topics" 
        WHERE name IN (
            'Child Screening and Assessment',
            'Teaching / Caregiving Practices',
            'Disabilities Services',
            'Ongoing Monitoring Management System',
            'Training and Professional Development'
        ) 
        AND "deletedAt" IS NULL;`,
```

**Review by GarrettEHill — APPROVED** [review]

File: src/migrations/20220812205424-new-ar-topics.js [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+        { transaction },
+      );
+
+      // Disable audit logging
+      await queryInterface.sequelize.query(
+        `
+          SELECT "ZAFSetTriggerState"(null, null, null, 'DISABLE');
+          `,
+        { transaction },
+      );
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:30]: Delete this, so audit log will record it. Per conversation Nathan & Garrett

File: src/migrations/20220812205424-new-ar-topics.js [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+        { transaction },
+      )));
+
+      // Enable audit logging
+      await queryInterface.sequelize.query(
+        `
+          SELECT "ZAFSetTriggerState"(null, null, null, 'ENABLE');
+          `,
+        { transaction },
+      );
```
Comment (GarrettEHill) [inline src/migrations/20220812205424-new-ar-topics.js:176]: This one too

Comment (nvms) [reply src/migrations/20220812205424-new-ar-topics.js:176]: The above looks like it enables logging. Do you mean instead for me to remove the bit that disables logging?

e.g. the below?

https://github.com/HHS/Head-Start-TTADP/blob/ef3e04df52a2cffc3fd3ea8fb912f84d5534a7c2/src/migrations/20220812205424-new-ar-topics.js#L112-L118

File: src/migrations/20220812205424-new-ar-topics.js (hunk 11 above)
Comment (nvms) [reply src/migrations/20220812205424-new-ar-topics.js]: Done: 0b4d25dc27e6f8fafab0fa233e995b3ec4fa4dff

File: src/migrations/20220812205424-new-ar-topics.js (hunk 12 above)
Comment (nvms) [reply src/migrations/20220812205424-new-ar-topics.js:160]: (you cleared this up for me in slack)

Done: 27d833626837cac70a21055f656c099716128eec

**Pull request merged**